src/Swarm/Game/WorldGen.hs

```diff
@@ -113,17 +113,47 @@ testWorld2FromArray arr seed co@(Coords (r, c))
   tw2 = testWorld2 seed
   bnds = bounds arr
 
+-- | Offset a world by a multiple of the @skip@ in such a way that it
+--   satisfies the given predicate.
+findOffset :: Integer -> (WorldFun t Text -> Bool) -> WorldFun t Text -> WorldFun t Text
+findOffset skip isGood f = f'
+ where
+  offset :: Enumeration Int64
+  offset = fromIntegral . (skip *) <$> int
+
+  f' =
+    fromMaybe (error "the impossible happened, no offsets were found!")
+      . find isGood
+      . map shift
+      . enumerate
+      $ offset >< offset
+
+  shift (dr, dc) (Coords (r, c)) = f (Coords (r - dr, c - dc))
+
+-- | Offset the world so the base starts in a 32x32 patch containing at least one
+--   of each of a list of required entities.
+findPatchWith :: [Text] -> WorldFun t Text -> WorldFun t Text
+findPatchWith reqs = findOffset 32 isGoodPatch
+ where
+  patchCoords = [(r, c) | r <- [-16 .. 16], c <- [-16 .. 16]]
+  isGoodPatch f = all (`S.member` es) reqs
+   where
+    es = S.fromList . mapMaybe (snd . f . Coords) $ patchCoords
+
 -- | Offset the world so the base starts on empty spot next to tree and grass.
-findGoodOrigin :: WorldFun t Text -> WorldFun t Text
-findGoodOrigin f = \(Coords (r, c)) -> f (Coords (r + rOffset, c + cOffset))
+findTreeOffset :: WorldFun t Text -> WorldFun t Text
+findTreeOffset = findOffset 1 isGoodPlace
```

Comment:
Which direction will this search?

I am concerned if it is guaranteed that the good place will stay in the good patch.
If it spirals out, then it should be fine, but if it goes in line, then it could escape. :thinking:

Reply:
Good question.  The order in which it searches comes from the way the `simple-enumeration` package works.  `Int` values are enumerated in the order `0, 1, -1, 2, -2, 3, -3 ...`, and then pairs of ints are enumerated using a balanced diagonal sort of scheme.  It's not exactly a spiral, but generally speaking closer locations are checked before farther ones:

```
λ> take 20 . enumerate $ int >< int
[(0,0),(0,1),(1,0),(0,-1),(1,1),(-1,0),(0,2),(1,-1),(-1,1),(2,0),(0,-2),(1,2),(-1,-1),(2,1),(-2,0),(0,3),(1,-2),(-1,2),(2,-1),(-2,1)]
```